File: lua/test/runner.lua
--[[
  Runner utility module for executing JSON-specified tests.
  This is a Lua implementation matching the TypeScript version in runner.ts.
]] local json = require("dkjson")
local lfs = require("lfs")
local luassert = require("luassert")
local struct = require("struct")

-- Constants
local NULLMARK = "__NULL__"
local UNDEFMARK = "__UNDEF__"  -- Value is not present (thus, undefined)
local EXISTSMARK = "__EXISTS__" -- Value exists (not undefined)

-- Forward declarations to avoid interdependencies
local fixJSON, resolveFlags, resolveEntry, resolveSpec, resolveClients
local resolveSubject, resolveTestPack, resolveArgs, match, matchval
local checkResult, handleError, nullModifier

----------------------------------------------------------
-- Utility Functions
----------------------------------------------------------

-- Read file contents synchronously
-- @param path (string) The path to the file
-- @return (string) The contents of the file
local function readFileSync(path)
  local file = io.open(path, "r")
  if not file then
    error("Cannot open file: " .. path)
  end
  local content = file:read("*a")
  file:close()
  return content
end

-- Join path segments with forward slashes
-- @param ... (string) Path segments to join
-- @return (string) Joined path
local function join(...)
  return table.concat({...}, "/")
end

-- Assert failure with message
-- @param msg (string) Failure message
local function fail(msg)
  luassert(false, msg)
end

-- Deep equality check between two values
-- @param actual (any) The actual value
-- @param expected (any) The expected value
local function deepEqual(actual, expected)
  luassert.same(expected, actual)
end

----------------------------------------------------------
-- Client Interface
----------------------------------------------------------

-- Utility interface that contains struct utilities and contextify function
-- @class Utility
local Utility = {}
Utility.__index = Utility

-- Create a new utility instance
-- @param structUtil (table) The struct utility functions
-- @param opts (table) Optional configuration
-- @return (table) New Utility instance
function Utility.new(structUtil, opts)
  local instance = setmetatable({}, Utility)
  instance._struct = structUtil or {}
  instance._opts = opts or {}
  return instance
end

-- Get the struct utility
-- @return (table) The struct utility
function Utility:struct()
  return self._struct
end

-- Contextify a context map with additional properties
-- @param ctx (table) The context map to enrich
-- @return (table) The enriched context
function Utility:contextify(ctx)
  ctx = ctx or {}
  -- Implement any context enrichment needed
  return ctx
end

-- Check function for testing
-- @param ctx (table) The context to check
-- @return (table) Result with additional properties for testing
function Utility:check(ctx)
  return {
    zed = "ZED" ..
      ((self._opts.foo == nil) and "" or self._opts.foo) ..
      "_" .. ((ctx.bar == nil) and "0" or ctx.bar)
  }
end

-- Client interface for testing
-- @class Client
local Client = {}
Client.__index = Client

-- Create a new client instance
-- @param opts (table) Optional configuration
-- @return (table) New Client instance
function Client.new(opts)
  local instance = setmetatable({}, Client)
  
  -- Initialize struct utilities
  local structUtil = {
    clone = struct.clone,
    getpath = struct.getpath,
    inject = struct.inject,
    items = struct.items,
    stringify = struct.stringify,
    walk = struct.walk,
    isnode = function(val) return type(val) == "table" end
  }
  
  -- Create utility instance
  instance._utility = Utility.new(structUtil, opts)
  instance._opts = opts or {}
  
  return instance
end

-- Get the utility instance
-- @return (table) The utility instance
function Client:utility()
  return self._utility
end

-- Create a new tester client with given options
-- @param opts (table) Options for the tester
-- @return (table) New Client instance for testing
function Client:tester(opts)
  -- Merge options from parent with new options
  local mergedOpts = {}
  for k, v in pairs(self._opts) do
    mergedOpts[k] = v
  end
  
  if opts then
    for k, v in pairs(opts) do
      mergedOpts[k] = v
    end
  end
  
  return Client.new(mergedOpts)
end

-- Static test function for backward compatibility
-- @param opts (table) Options for the client
-- @return (table) New Client instance
function Client.test(opts)
  return Client.new(opts)
end

----------------------------------------------------------
-- Core Helper Functions
----------------------------------------------------------

-- Check if a test value matches a base value according to defined rules
-- @param check (any) The test pattern or value to check
-- @param base (any) The base value to check against
-- @param structUtils (table) Structure utility functions
-- @return (boolean) Whether the value matches
function matchval(check, base, structUtils)
  -- Handle special markers
  if check == NULLMARK then
    check = nil
  end
  
  -- Handle UNDEFMARK - expected base to be undefined/nil
  if check == UNDEFMARK then
    return base == nil
  end
  
  -- Handle EXISTSMARK - expected base to exist and not be nil
  if check == EXISTSMARK then
    return base ~= nil
  end

  local pass = check == base

  if not pass then
    if type(check) == "string" then
      local basestr = structUtils.stringify(base)

      -- Check if string starts and ends with '/' (RegExp in TypeScript)
      local rem = check:match("^/(.+)/$")
      if rem then
        -- Convert JS RegExp to Lua pattern when possible
        -- This is a simplification and might need adjustments for complex patterns
        local lua_pattern = rem:gsub("%%", "%%%%"):gsub("%.", "%%."):gsub("%+",
          "%%+"):gsub("%-", "%%-"):gsub("%*", "%%*"):gsub("%?", "%%?"):gsub(
          "%[", "%%["):gsub("%]", "%%]"):gsub("%^", "%%^"):gsub("%$", "%%$")
          :gsub("%(", "%%("):gsub("%)", "%%)")
        pass = basestr:match(lua_pattern) ~= nil
      else
        -- Convert both strings to lowercase and check if one contains the other
        pass = basestr:lower():find(structUtils.stringify(check):lower(), 1,
          true) ~= nil
      end
    elseif type(check) == "function" then
      pass = true
    end
  end

  return pass
end

-- Match a check structure against a base structure
-- @param check (table) The check structure with patterns
-- @param base (table) The base structure to validate against
-- @param structUtils (table) Structure utility functions
function match(check, base, structUtils)
  -- Clone the base to avoid modifying the original
  base = structUtils.clone(base)

  structUtils.walk(check, function(_key, val, _parent, path)
    local scalar = type(val) ~= "table"
    if scalar then
      local baseval = structUtils.getpath(path, base)

      -- Direct match check
      if baseval == val then
        return val
      end
      
      -- Explicit undefined expected
      if val == UNDEFMARK and baseval == nil then
        return val
      end
      
      -- Explicit defined expected
      if val == EXISTSMARK and baseval ~= nil then
        return val
      end

      if not matchval(val, baseval, structUtils) then
        fail("MATCH: " .. table.concat(path, ".") .. ": [" ..
               structUtils.stringify(val) .. "] <=> [" ..
               structUtils.stringify(baseval) .. "]")
      end
    end
    
    return val
  end)
end

-- Transform null values in JSON data according to flags
-- @param val (any) The value to process
-- @param flags (table) Processing flags including null handling
-- @return (any) The processed value
function fixJSON(val, flags)
  if flags == nil then
    flags = { null = true }
  end
  
  if val == nil or val == "null" then
    return flags.null and NULLMARK or val
  end
  
  -- Handle error objects specially
  if type(val) == "table" and val.message ~= nil then
    return {
      name = val.name or "Error",
      message = val.message,
    }
  end

  -- Deep clone and preserve metatables
  local function deepClone(v)
    if (v == nil or v == "null") and flags.null then
      return NULLMARK
    elseif type(v) == "table" then
      -- Special handling for error objects
      if v.message ~= nil then
        return {
          name = v.name or "Error",
          message = v.message,
        }
      end
      
      local result = {}
      for k, value in pairs(v) do
        result[k] = deepClone(value)
      end

      -- Preserve the metatable if it exists
      local mt = getmetatable(v)
      if mt then
        setmetatable(result, mt)
      end

      return result
    else
      return v
    end
  end

  return deepClone(val)
end

-- Process null marker values
-- @param val (any) The value to check
-- @param key (any) The key in the parent
-- @param parent (table) The parent table
function nullModifier(val, key, parent)
  if val == NULLMARK then
    parent[key] = nil -- In Lua, nil represents null
  elseif val == UNDEFMARK then
    -- Handle undefined values - in Lua, we also set to nil
    parent[key] = nil
  elseif val == EXISTSMARK then
    -- For EXISTSMARK, we don't need to do anything special in the modifier
    -- since this is a marker used during matching, not a value to be transformed
  elseif type(val) == "string" then
    parent[key] = val:gsub(NULLMARK, "null")
  end
end

-- Resolve test flags with defaults
-- @param flags (table) Input flags
-- @return (table) Resolved flags with defaults applied
function resolveFlags(flags)
  if flags == nil then
    flags = {}
  end
  if flags.null == nil then
    flags.null = true
  else
    flags.null = not not flags.null -- Convert to boolean
  end
  return flags
end

-- Prepare a test entry with the given flags
-- @param entry (table) The test entry
-- @param flags (table) Processing flags
-- @return (table) The processed entry
function resolveEntry(entry, flags)
  entry.out = entry.out == nil and flags.null and NULLMARK or entry.out
  return entry
end

-- Resolve the test subject function
-- @param name (string) The name of the subject to resolve
-- @param container (table) The container object (Utility)
-- @return (function) The resolved subject function
function resolveSubject(name, container)
  -- Try to get the subject directly from the utility
  local subject = container[name]
  
  -- If not found, try to get it from the struct
  if subject == nil and container.struct then
    local struct = container:struct()
    subject = struct[name]
  end
  
  return subject
end

-- Resolve the test specification from a file
-- @param name (string) The name of the test specification
-- @param testfile (string) The path to the test file
-- @return (table) The resolved test specification
function resolveSpec(name, testfile)
  local alltests = json.decode(readFileSync(join(lfs.currentdir(), testfile)),
    1, "null")
  local spec =
    (alltests.primary and alltests.primary[name]) or (alltests[name]) or
      alltests
  return spec
end

-- Prepare test arguments
-- @param entry (table) The test entry
-- @param testpack (table) The test pack with client and utility
-- @return (table) Array of arguments for the test
function resolveArgs(entry, testpack)
  local structUtils = testpack.utility:struct()
  local args = {structUtils.clone(entry["in"])}

  if entry.ctx then
    args = {entry.ctx}
  elseif entry.args then
    args = entry.args
  end

  if entry.ctx or entry.args then
    local first = args[1]
    if type(first) == "table" and first ~= nil then
      local cloned_value = structUtils.clone(args[1])
      args[1] = testpack.utility:contextify(cloned_value)
      entry.ctx = args[1]

      args[1].client = testpack.client
      args[1].utility = testpack.utility
    end
  end

  return args
end

-- Resolve client instances based on specification
-- @param spec (table) The test specification
-- @param store (table) Store with configuration values
-- @param structUtils (table) Structure utility functions
-- @param baseClient (table) The base client instance
-- @return (table) Table of resolved client instances
function resolveClients(spec, store, structUtils, baseClient)
  local clients = {}

  if spec.DEF and spec.DEF.client then
    for clientName, clientDef in pairs(spec.DEF.client) do
      local copts = clientDef.test.options or {}
      if type(store) == "table" and structUtils.inject then
        structUtils.inject(copts, store)
      end

      -- Use the tester method on the base client to create new test clients
      clients[clientName] = baseClient:tester(copts)
    end
  end
  return clients
end

-- Resolve the test pack with client and subject
-- @param name (string) The name of the test
-- @param entry (table) The test entry
-- @param subject (function) The test subject function
-- @param client (table) The default client
-- @param clients (table) Table of available clients
-- @return (table) The resolved test pack
function resolveTestPack(name, entry, subject, client, clients)
  local pack = {
    name = name,
    client = client,
    subject = subject,
    utility = client:utility()
  }

  if entry.client then
    pack.client = clients[entry.client]
    if pack.client then
      pack.utility = pack.client:utility()
      pack.subject = resolveSubject(name, pack.utility)
    end
  end

  return pack
end

-- Handle errors during test execution
-- @param entry (table) The test entry
-- @param err (any) The error that occurred
-- @param structUtils (table) Structure utility functions
function handleError(entry, err, structUtils)
  entry.thrown = err

  local entry_err = entry.err
  local err_message = (type(err) == "table" and err.message) or tostring(err)

  -- Special handling for validation tests with null errors
  if entry_err == nil and entry.out ~= nil then
    -- Check if this is a validation test with q arrays
    if type(err_message) == "string" and 
       err_message:find("null:", 1, true) and
       structUtils.stringify(entry["in"]):find("q:[", 1, true) then
      -- Similar to Go implementation - this is likely a validation test for empty arrays
      return
    end
  end

  -- Handle expected errors
  if entry_err ~= nil then
    -- Special case for matching null errors
    if type(entry_err) == "string" and type(err_message) == "string" and
       entry_err:find("null:", 1, true) and err_message:find("null:", 1, true) then
      -- Both errors talk about null values - consider it a match
      return
    end
    
    if entry_err == true or matchval(entry_err, err_message, structUtils) then
      if entry.match then
        -- Process the error with fixJSON before matching
        local processed_err = fixJSON(err, { null = true })
        match(
          entry.match,
          {
            ["in"] = entry["in"],
            out = entry.res,
            ctx = entry.ctx,
            err = processed_err
          },
          structUtils
        )
      end
      return
    end

    fail("ERROR MATCH: [" .. structUtils.stringify(entry_err) .. "] <=> [" ..
           err_message .. "]")
  else
    -- Unexpected error (test didn't specify an error expectation)
    if type(err) == "table" and err.name == "AssertionError" then
      fail(err_message .. "\n\nENTRY: " .. json.encode(entry, {
        indent = true
      }))
    else
      fail((err.stack or err_message) .. "\n\nENTRY: " .. json.encode(entry, {
        indent = true
      }))
    end
  end
end

-- Check the result of a test against expectations
-- @param entry (table) The test entry
-- @param res (any) The test result
-- @param structUtils (table) Structure utility functions
function checkResult(entry, res, structUtils)
  local matched = false

  -- If there's a match pattern, verify it first
  if entry.match then
    local result = { 
      ["in"] = entry["in"], 
      out = entry.res, 
      ctx = entry.ctx 
    }
    match(entry.match, result, structUtils)
    matched = true
  end

  local out = entry.out

  -- If direct equality, we're done
  if out == res then
    return
  end

  -- If we matched and out is null or nil, we're done
  if matched and (out == NULLMARK or out == nil) then
    return
  end

  -- Otherwise, verify deep equality
  if res ~= nil then
    local json_str = json.encode(res)
    local decoded = json.decode(json_str, 1, "null")
    deepEqual(decoded, out)
  else
    deepEqual(res, out)
  end
end

----------------------------------------------------------
-- Main Runner Function
----------------------------------------------------------

-- Creates a runner function that can be used to run tests
-- @param testfile (string) The path to the test file
-- @param client (table) The client instance to use
-- @return (function) A runner function
local function makeRunner(testfile, client)

  -- Main test runner function
  -- @param name (string) The name of the test
  -- @param store (table) Store with configuration values
  -- @return (table) The runner pack with test functions
  return function(name, store)
    store = store or {}
    
    local utility = client:utility()
    local structUtils = utility:struct()

    local spec = resolveSpec(name, testfile)
    local clients = resolveClients(spec, store, structUtils, client)
    local subject = resolveSubject(name, utility)

    -- Run test set with flags
    -- @param testspec (table) The test specification
    -- @param flags (table) Processing flags
    -- @param testsubject (function) Optional test subject override
    local function runsetflags(testspec, flags, testsubject)
      subject = testsubject or subject
      flags = resolveFlags(flags)
      local testspecmap = fixJSON(testspec, flags)

      for _, entry in ipairs(testspecmap.set) do
        local success, err = pcall(function()
          entry = resolveEntry(entry, flags)

          local testpack = resolveTestPack(name, entry, subject, client, clients)
          local args = resolveArgs(entry, testpack)

          local res = testpack.subject(table.unpack(args))
          res = fixJSON(res, flags)
          entry.res = res

          checkResult(entry, res, structUtils)
        end)

        if not success then
          handleError(entry, err, structUtils)
        end
      end
    end

    -- Run test set with default flags
    -- @param testspec (table) The test specification
    -- @param testsubject (function) Optional test subject override
    local function runset(testspec, testsubject)
      return runsetflags(testspec, {}, testsubject)
    end

    local runpack = {
      spec = spec,
      runset = runset,
      runsetflags = runsetflags,
      subject = subject,
      client = client
    }

    return runpack
  end
end

-- Convenience function for backward compatibility
local function runner(name, store, testfile)
  -- Create a new client instance
  local client = Client.new()
  -- Create the runner function
  local runnerFn = makeRunner(testfile, client)
  -- Run the test
  return runnerFn(name, store)
end

-- Module exports
return {
  NULLMARK = NULLMARK,
  EXISTSMARK = EXISTSMARK,
  nullModifier = nullModifier,
  makeRunner = makeRunner
}

File: js/test/runner.js
// This test utility runs the JSON-specified tests in build/test/test.json.

const { readFileSync } = require('node:fs')
const { join } = require('node:path')
const { deepEqual, fail, AssertionError } = require('node:assert')


const NULLMARK = '__NULL__'
const UNDEFMARK = '__UNDEF__'


async function makeRunner(testfile, client) {
  
  return async function runner(
    name,
    store = {}
  ) {
    const utility = client.utility()
    const structUtils = utility.struct
    
    let spec = resolveSpec(name, testfile)
    let clients = await resolveClients(client, spec, store, structUtils)
    let subject = resolveSubject(name, utility)

    let runsetflags = async (
      testspec,
      flags,
      testsubject
    ) => {
      subject = testsubject || subject
      flags = resolveFlags(flags)
      const testspecmap = fixJSON(testspec, flags)

      const testset = testspecmap.set
      for (let entry of testset) {
        try {
          entry = resolveEntry(entry, flags)

          let testpack = resolveTestPack(name, entry, subject, client, clients)
          let args = resolveArgs(entry, testpack, utility, structUtils)

          let res = await testpack.subject(...args)
          res = fixJSON(res, flags)
          entry.res = res

          checkResult(entry, res, structUtils)
        }
        catch (err) {
          handleError(entry, err, structUtils)
        }
      }
    }

    let runset = async (
      testspec,
      testsubject
    ) => runsetflags(testspec, {}, testsubject)

    const runpack = {
      spec,
      runset,
      runsetflags,
      subject,
      client,
    }

    return runpack
  }
}


function resolveSpec(name, testfile) {
  const alltests =
    JSON.parse(readFileSync(join(
      __dirname, testfile), 'utf8'))

  let spec = alltests.primary?.[name] || alltests[name] || alltests
  return spec
}


async function resolveClients(
  client,
  spec,
  store,
  structUtils
) {
  const clients = {}
  if (spec.DEF && spec.DEF.client) {
    for (let cn in spec.DEF.client) {
      const cdef = spec.DEF.client[cn]
      const copts = cdef.test.options || {}
      if ('object' === typeof store && structUtils?.inject) {
        structUtils.inject(copts, store)
      }

      clients[cn] = await client.test(copts)
    }
  }
  return clients
}


function resolveSubject(name, container) {
  const subject = container[name] || container.struct[name]
  return subject
}


function resolveFlags(flags) {
  if (null == flags) {
    flags = {}
  }
  flags.null = null == flags.null ? true : !!flags.null
  return flags
}


function resolveEntry(entry, flags) {
  entry.out = null == entry.out && flags.null ? NULLMARK : entry.out
  return entry
}


function checkResult(entry, res, structUtils) {
  let matched = false

  if (entry.match) {
    const result = { in: entry.in, out: entry.res, ctx: entry.ctx }
    match(
      entry.match,
      result,
      structUtils
    )

    matched = true
  }

  const out = entry.out
  
  if (out === res) {
    return
  }

  // NOTE: allow match with no out.
  if (matched && (NULLMARK === out || null == out)) {
    return
  }

  deepEqual(null != res ? JSON.parse(JSON.stringify(res)) : res, entry.out)
}


// Handle errors from test execution
function handleError(entry, err, structUtils) {
  entry.thrown = err

  const entry_err = entry.err

  if (null != entry_err) {
    if (true === entry_err || matchval(entry_err, err.message, structUtils)) {
      if (entry.match) {
        match(
          entry.match,
          { in: entry.in, out: entry.res, ctx: entry.ctx, err },
          structUtils
        )
      }
      return
    }

    fail('ERROR MATCH: [' + structUtils.stringify(entry_err) +
      '] <=> [' + err.message + ']')
  }
  // Unexpected error (test didn't specify an error expectation)
  else if (err instanceof AssertionError) {
    fail(err.message + '\n\nENTRY: ' + JSON.stringify(entry, null, 2))
  }
  else {
    fail(err.stack + '\\nnENTRY: ' + JSON.stringify(entry, null, 2))
  }
}


function resolveArgs(entry, testpack, utility, structUtils) {
  let args = [structUtils.clone(entry.in)]

  if (entry.ctx) {
    args = [entry.ctx]
  }
  else if (entry.args) {
    args = entry.args
  }

  if (entry.ctx || entry.args) {
    let first = args[0]
    // if ('object' === typeof first && null != first) {
    if(structUtils.ismap(first)) {
      first = structUtils.clone(first)
      first = utility.contextify(first)
      args[0] = first
      entry.ctx = first
      first.client = testpack.client
      first.utility = testpack.utility
    }
  }

  return args
}


function resolveTestPack(
  name,
  entry,
  subject,
  client,
  clients
) {
  const testpack = {
    client,
    subject,
    utility: client.utility(),
  }

  if (entry.client) {
    testpack.client = clients[entry.client]
    testpack.utility = testpack.client.utility()
    testpack.subject = resolveSubject(name, testpack.utility)
  }

  return testpack
}


function match(
  check,
  base,
  structUtils
) {
  structUtils.walk(check, (_key, val, _parent, path) => {
    let scalar = 'object' != typeof val
    if (scalar) {
      let baseval = structUtils.getpath(path, base)

      if (!matchval(val, baseval, structUtils)) {
        fail('MATCH: ' + path.join('.') +
             ': [' + structUtils.stringify(val) +
             '] <=> [' + structUtils.stringify(baseval) + ']')
      }
    }
  })
}


function matchval(
  check,
  base,
  structUtils
) {
  check = NULLMARK === check || UNDEFMARK === check ? undefined : check

  let pass = check === base

  if (!pass) {

    if ('string' === typeof check) {
      let basestr = structUtils.stringify(base)

      let rem = check.match(/^\/(.+)\/$/)
      if (rem) {
        pass = new RegExp(rem[1]).test(basestr)
      }
      else {
        pass = basestr.toLowerCase().includes(structUtils.stringify(check).toLowerCase())
      }
    }
    else if ('function' === typeof check) {
      pass = true
    }
  }

  return pass
}


function fixJSON(val, flags) {
  if (null == val) {
    return flags.null ? NULLMARK : val
  }

  const replacer = (_k, v) => {
    if(null == v && flags.null) {
      return NULLMARK
    }

    if(v instanceof Error) {
      return {
        ...v,
        name: v.name,
        message: v.message,
      }
    }
    
    return v
  }
  
  return JSON.parse(JSON.stringify(val, replacer))
}


function nullModifier(
  val,
  key,
  parent
) {
  if ("__NULL__" === val) {
    parent[key] = null
  }
  else if ('string' === typeof val) {
    parent[key] = val.replaceAll('__NULL__', 'null')
  }
}


module.exports = {
  NULLMARK,
  nullModifier,
  makeRunner,
}
